File: linked-lists/Cpp_C/design_linked_list.c
#include<stdlib.h>

typedef struct node {
    int val;
    struct node* next;
    struct node* prev;
} node;


typedef struct {
    node* right;
    node* left;
} MyLinkedList;

node* createNode(int val) {
    node* nd = (node*)malloc(sizeof(node));
    nd->val = val;
    nd->next = NULL;
    nd->prev = NULL;
    return nd;
}

MyLinkedList* myLinkedListCreate() {
    MyLinkedList* obj = (MyLinkedList*)malloc(sizeof(MyLinkedList));
    obj->right = createNode(0);
    obj->left = createNode(0);
    obj->right->prev = obj->left;
    obj->left->next = obj->right;
    return obj;
}

int myLinkedListGet(MyLinkedList* obj, int index) {
    node* cur = obj->left->next;
    while (cur && index > 0) {
        cur = cur->next;
        index--;
    }
    if (cur && cur != obj->right && index == 0) {
        return cur->val;
    }
    return -1;
}

void myLinkedListAddAtHead(MyLinkedList* obj, int val) {
    node* Node = createNode(val);
    node* pre = obj->left;
    node* nxt = obj->left->next;

    pre->next = Node;
    Node->prev = pre;
    Node->next = nxt;
    nxt->prev = Node;
}

void myLinkedListAddAtTail(MyLinkedList* obj, int val) {
    node* Node = createNode(val);
    node* pre = obj->right->prev;
    node* nxt = obj->right;

    pre->next = Node;
    Node->prev = pre;
    Node->next = nxt;
    nxt->prev = Node;

}

void myLinkedListAddAtIndex(MyLinkedList* obj, int index, int val) {
    node* cur = obj->left->next;
    while (cur && index > 0) {
        cur = cur->next;
        index--;
    }
    if (cur && cur != obj->right->next && index == 0) {
        node* Node = createNode(val);
        node* pre = cur->prev;
        node* nxt = cur;

        pre->next = Node;
        Node->prev = pre;
        Node->next = nxt;
        nxt->prev = Node;
    }
}

void myLinkedListDeleteAtIndex(MyLinkedList* obj, int index) {
    node* cur = obj->left->next;
    while(cur && index > 0) {
        cur = cur->next;
        index--;
    }
    if (cur && cur != obj->right && index == 0){
        node* pre = cur->prev;
        node* nxt = cur->next;

        pre->next = nxt;
        nxt->prev = pre;
    }
}

void myLinkedListFree(MyLinkedList* obj) {
    free(obj->right);
    free(obj->left);
    free(obj);
}

File: linked-lists/Python/design_linked_list.py
class ll:

    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None


class MyLinkedList:

    def __init__(self):
        self.left = ll(0)
        self.right = ll(0)
        self.left.next = self.right
        self.right.prev = self.left

    def get(self, index: int) -> int:
        cur = self.left.next
        while cur and index > 0:
            cur = cur.next
            index -= 1
        if cur and cur != self.right and index == 0:
            return cur.val
        return -1

    def addAtHead(self, val: int) -> None:
        node, pre, nxt = ll(val), self.left, self.left.next
        pre.next = node
        node.prev = pre
        nxt.prev = node
        node.next = nxt

    def addAtTail(self, val: int) -> None:
        node, pre, nxt = ll(val), self.right.prev, self.right
        pre.next = node
        node.prev = pre
        nxt.prev = node
        node.next = nxt

    def addAtIndex(self, index: int, val: int) -> None:
        cur = self.left.next
        while cur and index > 0:
            cur = cur.next
            index -= 1
        if cur and index == 0:
            node, pre, nxt = ll(val), cur.prev, cur
            pre.next = node
            node.prev = pre
            nxt.prev = node
            node.next = nxt

    def deleteAtIndex(self, index: int) -> None:
        cur = self.left.next
        while cur and index > 0:
            cur = cur.next
            index -= 1
        if cur and cur != self.right and index == 0:
            pre, nxt = cur.prev, cur.next
            pre.next = nxt
            nxt.prev = pre
